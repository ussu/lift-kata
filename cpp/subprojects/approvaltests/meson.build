project('approvaltests',
    'cpp',
    version : '7.0.0',
    license : 'Apache'
)

approvaltests_dep = declare_dependency(
    include_directories : include_directories('.')
)