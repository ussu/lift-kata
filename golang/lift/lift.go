package lift

// Direction ..
type Direction int

// Directions ..
const (
	Up Direction = iota
	Down
)

// Call ..
type Call struct {
	Floor     int
	Direction Direction
}

// Lift ..
type Lift struct {
	ID        string
	Floor     int
	Requests  []int
	DoorsOpen bool
}

// System ..
type System struct {
	floors []int
	lifts  []Lift
	calls  []Call
}

// NewSystem ..
func NewSystem() *System {
	return &System{floors: []int{}, lifts: []Lift{}, calls: []Call{}}
}

// AddFloors ..
func (s *System) AddFloors(floors ...int) {
	s.floors = append(s.floors, floors...)
}

// AddLifts ..
func (s *System) AddLifts(lifts ...Lift) {
	s.lifts = append(s.lifts, lifts...)
}

// AddCalls ..
func (s *System) AddCalls(calls ...Call) {
	s.calls = append(s.calls, calls...)
}

// CallsFor ..
func (s System) CallsFor(floor int) (calls []Call) {
	calls = []Call{}
	for _, c := range s.calls {
		if c.Floor == floor {
			calls = append(calls, c)
		}
	}
	return calls
}

// Tick ..
func (s System) Tick() {
	panic("Implement this method")
}
