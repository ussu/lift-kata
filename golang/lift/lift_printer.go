package lift

import (
	"fmt"
	"strconv"
	"strings"
)

// Printer ..
type Printer interface{
	PrintLift(lift Lift, floor int) string
}

type simplePrinter struct {}

// NewSimplePrinter ..
func NewSimplePrinter() Printer {
	return 	&simplePrinter{}
}

// PrintLift ..
func (p simplePrinter) PrintLift(lift Lift, floor int) (liftStr string) {
	if inRequestedFloor(lift, floor){
		liftStr = fmt.Sprintf(" *%s ", lift.ID)
	} else {
		liftStr = fmt.Sprintf("  %s ", lift.ID)
	}
	return liftStr
}

type printer struct {}

// NewPrinter ..
func NewPrinter() Printer {
	return &printer{}
}

// PrintLift ..
func (p printer) PrintLift(lift Lift, floor int) (liftStr string) {
	if lift.DoorsOpen {
		if inRequestedFloor(lift, floor) {
			liftStr = fmt.Sprintf("]*%s[", lift.ID)
		} else {
			liftStr = fmt.Sprintf(" ]%s[", lift.ID)
		}
	
	} else {
		if inRequestedFloor(lift, floor){
			liftStr = fmt.Sprintf("[*%s]", lift.ID)
		} else {
			liftStr = fmt.Sprintf(" [%s]", lift.ID)
		}
	}
	return liftStr
}

// PrintLifts ...
func PrintLifts(liftSystem *System, liftPrinter Printer) string {
	result := ""
	floorNumberLength := calculateFloorNumberLength(liftSystem.floors)
	for _, floor := range reverseLiftFloors(liftSystem.floors) {
		calls := printCalls(liftSystem, floor)
		callPadding := whiteSpace(2 - len(calls))
		floorPadding := whiteSpace(floorNumberLength - len(strconv.Itoa(floor)))
		lifts := printLiftsForFloor(liftSystem, liftPrinter, floor)
		result += fmt.Sprintf("%s%d %s%s %s %s%d\n", 
		floorPadding, floor, strings.Join(calls, ""), callPadding, 
		strings.Join(lifts, " "), floorPadding, floor)
	}
	return result
}

func printLiftsForFloor(liftSystem *System, liftPrinter Printer, floor int) []string {
	lifts := []string{}
	for _, lift := range liftSystem.lifts {
		lifts = append(lifts, printLiftForFloor(liftPrinter, lift, floor))
	}
	return lifts
}

func printLiftForFloor(liftPrinter Printer, lift Lift, floor int) (liftStr string) {
	if lift.Floor == floor {
		liftStr = liftPrinter.PrintLift(lift, floor)
	} else {
		liftIDPadding := whiteSpace(len(lift.ID))
		if inRequestedFloor(lift, floor) {
			liftStr = fmt.Sprintf("  *%s", liftIDPadding)
		} else {
			liftStr = fmt.Sprintf("   %s", liftIDPadding)
		}
	}
	return liftStr
}

func inRequestedFloor(lift Lift, floor int) (found bool) {
	for _, request := range lift.Requests {
		if request == floor {
			found = true
		}	
	}
	return found
}

func printCalls(liftSystem *System, floor int) []string {
	calls := []string{}
	for _, call := range liftSystem.CallsFor(floor) {
		calls = append(calls, printCallDirection(call))
	}
	return calls
}

func printCallDirection(call Call) string {
	if call.Direction == Down {
		return "v"
	} else if call.Direction == Up {
		return "^"
	}
	return " "
}

func calculateFloorNumberLength(floors []int) int {
	if len(floors) < 1 {
		panic("Must have at least one floor")
	}
	lowestFloor := minInt(floors...)
	highestFloor := maxInt(floors...)
	longestFloorName := maxStringLength(strconv.Itoa(lowestFloor), strconv.Itoa(highestFloor))
	return len(longestFloorName)
}

func minInt(ints ...int) int {
	lowest := ints[0]
	for _, i := range ints {
		if i < lowest {
			lowest = i
		}
	}
	return lowest
}

func maxInt(ints ...int) int {
	highest := ints[0]
	for _, i := range ints {
		if i > highest {
			highest = i
		}
	}
	return highest
}

func maxStringLength(strings ...string) string {
	highest := strings[0]
	for _, s := range strings {
		if len(s) > len(highest) {
			highest = s
		}
	}
	return highest
}

func reverseLiftFloors(floors []int) []int {
	for i := len(floors)/2 - 1; i >= 0; i-- {
		opp := len(floors) - 1 - i
		floors[i], floors[opp] = floors[opp], floors[i]
	}
	return floors
}

func whiteSpace(length int) string {
	return strings.Repeat(" ", length)
}
