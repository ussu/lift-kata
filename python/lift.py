from enum import Enum

from attr import dataclass


class Direction(Enum):
    UP = 0
    DOWN = 1


@dataclass
class Call:
    floor: int
    direction: Direction


class Lift:

    def __init__(self, id, floor, doors_open=False, requested_floors=None):
        self.id = id
        self.floor = floor
        self.doors_open = doors_open
        self.requested_floors = set(requested_floors) if requested_floors else set()


class LiftSystem:
    def __init__(self, floors=None, lifts=None, calls=None):
        self.floors = list(floors) if floors else []
        self.lifts = list(lifts) if lifts else []
        self.calls = list(calls) if calls else []

    def calls_for(self, floor):
        return [c for c in self.calls if c.floor == floor]

    def tick(self):
        # TODO: implement this method
        pass


