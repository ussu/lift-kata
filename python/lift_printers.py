from lift import Direction


class SimpleLiftPrinter:
    def print_lift_for_floor(self, lift, floor):
        "Print information about a lift for a particular floor, including the position of the lift and requested floors."
        if lift.floor == floor:
            lift_str = self.print_lift(lift, floor)
        else:
            lift_id_padding = _whitespace(len(lift.id))
            if floor in lift.requested_floors:
                lift_str = f"  *{lift_id_padding}"
            else:
                lift_str = f"   {lift_id_padding}"
        return lift_str

    def print_lift(self, lift, floor):
        "Print information about a lift"
        if floor in lift.requested_floors:
            return f" *{lift.id} "
        else:
            return f"  {lift.id} "


class LiftPrinter(SimpleLiftPrinter):

    def print_lift(self, lift, floor):
        "Print information about a lift, including door status"
        if lift.doors_open:
            if floor in lift.requested_floors:
                return f"]*{lift.id}["
            else:
                return f" ]{lift.id}["

        else:
            if floor in lift.requested_floors:
                return f"[*{lift.id}]"
            else:
                return f" [{lift.id}]"


def print_lifts(lift_system, lift_printer=None):
    """
    Prints the state of a LiftSystem using ASCII art.
    """
    lift_printer = lift_printer or LiftPrinter()
    r = ""
    floor_number_length = _calculate_floor_number_length(lift_system.floors)
    for floor in reversed(lift_system.floors):
        calls = "".join([print_call_direction(call) for call in (lift_system.calls_for(floor))])
        # if there are less than 2 calls on a floor we add padding to keep everything aligned
        call_padding = _whitespace(2 - len(calls))

        lifts = " ".join([lift_printer.print_lift_for_floor(lift, floor) for lift in lift_system.lifts])

        # if the floor number doesn't use all the characters, pad with whitespace
        floor_padding = _whitespace(floor_number_length - len(str(floor)))

        # put the floor number at both ends of the line to make it more readable when there are lots of lifts,
        # and to prevent the IDE from doing rstrip on save and messing up the approved files.
        r += f"{floor_padding}{floor} {calls}{call_padding} {lifts} {floor_padding}{floor}\n"

    return r

def _calculate_floor_number_length(floors):
    if not floors:
        raise ValueError("Must have at least one floor")
    lowest_floor = min(floors)
    highest_floor = max(floors)
    longest_floor_name = max(str(lowest_floor), str(highest_floor))
    return len(longest_floor_name)

def print_lift_for_floor(lift, floor):
    "Print information about a lift for a particular floor, including the position of the lift and requested floors."
    if lift.floor == floor:
        lift_str = print_lift(lift, floor)
    else:
        padding = _whitespace(len(lift.id))
        if floor in lift.requested_floors:
            lift_str = f"  *{padding}"
        else:
            lift_str = f"   {padding}"
    return lift_str


def print_lift(lift, floor):
    "Print information about a lift, including door status"
    if lift.doors_open:
        if floor in lift.requested_floors:
            return f"]*{lift.id}["
        else:
            return f" ]{lift.id}["

    else:
        if floor in lift.requested_floors:
            return f"[*{lift.id}]"
        else:
            return f" [{lift.id}]"


def print_call_direction(call):
    "Print information about the direction of a Call"
    if call.direction == Direction.DOWN:
        return "v"
    elif call.direction == Direction.UP:
        return "^"
    else:
        # should be unreachable
        return " "


def _whitespace(length):
    "Return a string of whitespace with the requested length"
    return " "*length
