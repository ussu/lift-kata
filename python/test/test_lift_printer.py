from approvaltests import verify

from lift import LiftSystem, Lift, Direction, Call
from lift_printers import print_lifts, SimpleLiftPrinter


def test_no_lifts():
    lift_system = LiftSystem(floors=[0, 1, 2, 3])
    verify(print_lifts(lift_system))

def test_no_doors():
    liftA = Lift("A", floor=3, requested_floors=[0])
    lift_system = LiftSystem(floors=[0, 1, 2, 3],
                             lifts=[liftA],
                             calls=[Call(1, Direction.DOWN)])
    verify(print_lifts(lift_system, lift_printer=SimpleLiftPrinter()))


def test_sample_lift_system():
    liftA = Lift("A", floor=3, requested_floors=[0])
    liftB = Lift("B", floor=2)
    liftC = Lift("C", floor=2, doors_open=True)
    liftD = Lift("D", floor=0, requested_floors=[0])
    lift_system = LiftSystem(floors=[0, 1, 2, 3],
                             lifts=[liftA, liftB, liftC, liftD],
                             calls=[Call(1, Direction.DOWN)])
    verify(print_lifts(lift_system))


def test_illegal_states():
    liftA = Lift("A", floor=0, requested_floors=[0], doors_open=True)
    lifts = LiftSystem(floors=[0, 1], lifts=[liftA])
    verify(print_lifts(lifts))


def test_large_lift_system():
    liftA = Lift("A", floor=0, requested_floors=[3,5,7])
    liftB = Lift("B", floor=2, doors_open=True)
    liftC = Lift("C", floor=-2)
    liftC.requested_floors = [-2, 0]
    liftD = Lift("D", floor=8, doors_open=True)
    liftD.requested_floors = [0, -1, -2]
    liftSVC = Lift("SVC", floor=10)
    liftSVC.requested_floors = [0,-1]
    liftF = Lift("F", floor=8)
    lift_system = LiftSystem(floors=[-2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
                             lifts=[liftA, liftB, liftC, liftD, liftSVC, liftF],
                             calls=[Call(1, Direction.DOWN), Call(6, Direction.DOWN), Call(5, Direction.UP), Call(5, Direction.DOWN), Call(-1, Direction.UP)])
    verify(print_lifts(lift_system))

