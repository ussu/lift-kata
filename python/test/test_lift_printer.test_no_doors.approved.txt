3      A  3
2         2
1 v       1
0      *  0
